contracts/utils/PredictoorHelper.sol

```diff
@@ -3,36 +3,122 @@ pragma solidity 0.8.12;
+        consumeMarketFee _consumeMarketFee;
     }
+    struct FreParams {
+        address exchangeContract;
+        bytes32 exchangeId;
+        uint256 maxBaseTokenAmount;
+        uint256 swapMarketFee;
+        address marketFeeAddress;
+    }
+}
+
+contract PredictoorHelper is Ownable {
+    event OrderStarted(
+        address indexed consumer,
+        address payer,
+        uint256 amount,
+        uint256 serviceIndex,
+        uint256 timestamp,
+        address indexed publishMarketAddress,
+        uint256 blockNumber
+    );
+
+    function consumeMultiple(
+        address[] calldata addresses,
+        uint256[] calldata times,
+        address tokenAddress
+    ) external {
+        IERC20Template token = IERC20Template(tokenAddress);
+        uint256 balanceStart = token.balanceOf(address(this));
+        for (uint256 i = 0; i < addresses.length; i++) {
+            IERC20Template3.fixedRate[] memory frates = IERC20Template3(
+                addresses[i]
+            ).getFixedRates();
+            address fixedRateAddr = frates[0].contractAddress;
+            bytes32 id = frates[0].id;
+
+            IFixedRateExchange exchange = IFixedRateExchange(fixedRateAddr);
+
+            (uint256 baseTokenAmount, , , ) = exchange.calcBaseInGivenOutDT(
+                id,
+                1 ether,
+                0
+            );
+
+            require(
+                token.transferFrom(
+                    msg.sender,
+                    address(this),
+                    baseTokenAmount * times[i]
+                ),
+                "transfer failed"
+            );
 
-    modifier onlyAdmin() {
-        require(msg.sender == admin, "Not authorized");
-        _;
+            token.approve(addresses[i], baseTokenAmount * times[i]);
+            IERC20Template3.OrderParams memory orderParams = IERC20Template3
+                .OrderParams(
+                    msg.sender,
+                    0,
+                    IERC20Template.providerFee({
+                        providerFeeAddress: address(0),
+                        providerFeeToken: address(0),
+                        providerFeeAmount: 0,
+                        v: 0,
+                        r: bytes32(0),
+                        s: bytes32(0),
+                        validUntil: 0,
+                        providerData: ""
+                    }),
+                    IERC20Template.consumeMarketFee(address(0), address(0), 0)
+                );
+            IERC20Template3.FreParams memory freParams = IERC20Template3
+                .FreParams(fixedRateAddr, id, baseTokenAmount, 0, address(0));
+
+            for (uint256 j = 0; j < times[i]; j++) {
+                IERC20Template3(addresses[i]).buyFromFreAndOrder(
+                    orderParams,
+                    freParams
+                );
+            }
+        }
+        uint256 balanceAfter = token.balanceOf(address(this));
+        require(balanceAfter >= balanceStart, "lost balance");
```

Comment:
why do we have this condition?
In theory, at start, balanceStart should be 0.  then we transfer and pay for accesses.  if balanceAfter > balanceStart, we need to pay back some tokens